protocol/v1/qbft/instance/prepare.go

```diff
@@ -102,34 +100,48 @@ func (i *Instance) uponPrepareMsg() pipelines.SignedMessagePipeline {
 			return err
 		}
 
-		// TODO - calculate quorum one way (for prepare, commit, change round and decided) and refactor
-		if quorum, _ := i.containersMap[specqbft.PrepareMsgType].QuorumAchieved(signedMessage.Message.Round, prepareData.Data); quorum {
-			var errorPrp error
-			i.processPrepareQuorumOnce.Do(func() {
-				i.Logger.Info("prepared instance",
-					zap.String("identifier", hex.EncodeToString(i.State().GetIdentifier())), zap.Any("round", i.State().GetRound()))
-
-				// set prepared state
-				i.State().PreparedRound.Store(signedMessage.Message.Round)
-				i.State().PreparedValue.Store(prepareData.Data) // passing the data as is, and not get the specqbft.PrepareData cause of msgCount saves that way
-				i.ProcessStageChange(qbft.RoundStatePrepare)
-
-				// send commit msg
-				broadcastMsg, err := i.generateCommitMessage(i.State().GetPreparedValue())
-				if err != nil {
-					return
-				}
-				if e := i.SignAndBroadcast(broadcastMsg); e != nil {
-					i.Logger.Info("could not broadcast commit message", zap.Error(err))
-					errorPrp = e
-				}
-			})
-			return errorPrp
+		if quorum, _, _ := signedmsg.HasQuorum(i.ValidatorShare, i.containersMap[specqbft.PrepareMsgType].ReadOnlyMessagesByRound(i.State().GetRound())); !quorum {
+			return nil
+		}
+
+		if i.didSendCommitForHeightAndRound() {
```

Comment:
We doing that check with `processPrepareQuorumOnce`. 
i think let's comment this out and add todo to check if need to replace

Reply:
@nivBlox done